balancer/pickfirst/pickfirstleaf/pickfirstleaf_ext_test.go

```diff
@@ -851,6 +850,245 @@ func (s) TestPickFirstLeaf_EmptyAddressList(t *testing.T) {
+	// third server is contacted.
+	timerCh <- struct{}{}
+	if holds[2].Wait(ctx) != true {
+		t.Fatalf("Timeout waiting for server %d with address %q to be contacted", 2, addrs[2])
+	}
+
+	// First SubConn Fails.
+	holds[0].Fail(fmt.Errorf("test error"))
+
+	// No TF should be reported until the first pass is complete.
+	shortCtx, shortCancel := context.WithTimeout(ctx, defaultTestShortTimeout)
+	defer shortCancel()
+
+	testutils.AwaitNotState(shortCtx, t, cc, connectivity.TransientFailure)
+
+	// Third SubConn fails.
+	holds[2].Fail(fmt.Errorf("test error"))
+
+	testutils.AwaitNotState(shortCtx, t, cc, connectivity.TransientFailure)
+
+	// Last SubConn fails, this should result in a TF update.
+	holds[1].Fail(fmt.Errorf("test error"))
+	testutils.AwaitState(ctx, t, cc, connectivity.TransientFailure)
+}
+
+// TestPickFirstLeaf_HappyEyeballs_TriggerConnectionDelay verifies that
+// pickfirst attempts to connect to the second backend once the happy eyeballs
+// timer expires.
+func (s) TestPickFirstLeaf_HappyEyeballs_TriggerConnectionDelay(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	timerCh := make(chan struct{})
+	originalTimer := pfinternal.TimeAfterFunc
+	pfinternal.TimeAfterFunc = func(_ time.Duration, f func()) *time.Timer {
+		// Set a really long expiration to prevent it from triggering
+		// automatically.
+		timer := time.AfterFunc(time.Hour, f)
+		go func() {
+			select {
+			case <-ctx.Done():
+			case <-timerCh:
+			}
+			timer.Reset(0)
+		}()
+		return timer
+	}
+
+	defer func() {
+		pfinternal.TimeAfterFunc = originalTimer
+	}()
+
+	dialer := testutils.NewBlockingDialer()
+	opts := []grpc.DialOption{
+		grpc.WithDefaultServiceConfig(fmt.Sprintf(`{"loadBalancingConfig": [{"%s":{}}]}`, pickfirstleaf.Name)),
+		grpc.WithContextDialer(dialer.DialContext),
+	}
+	cc, rb, bm := setupPickFirstLeaf(t, 2, opts...)
+	addrs := bm.resolverAddrs()
+	holds := bm.holds(dialer)
+	rb.UpdateState(resolver.State{Addresses: addrs})
+	cc.Connect()
+
+	testutils.AwaitState(ctx, t, cc, connectivity.Connecting)
+
+	// Verify that the first server is contacted and the second is not.
+	if holds[0].Wait(ctx) != true {
+		t.Fatalf("Timeout waiting for server %d with address %q to be contacted", 0, addrs[0])
+	}
+	if holds[1].IsStarted() != false {
+		t.Fatalf("Server %d with address %q contacted unexpectedly", 1, addrs[1])
+	}
+
+	timerCh <- struct{}{}
+
+	// Second connection attempt is successful.
+	if holds[1].Wait(ctx) != true {
+		t.Fatalf("Timeout waiting for server %d with address %q to be contacted", 1, addrs[1])
+	}
+	holds[1].Resume()
+	testutils.AwaitState(ctx, t, cc, connectivity.Ready)
+}
+
+// TestPickFirstLeaf_HappyEyeballs_TFThenTimerFires tests the pickfirst balancer
+// by causing a SubConn to fail and then jumping to the 3rd SubConn after the
+// happy eyeballs timer expires.
+func (s) TestPickFirstLeaf_HappyEyeballs_TFThenTimerFires(t *testing.T) {
```

Comment:
Here and in the first test added in this PR, could we add an `_` after the `TF`? So, this would be `TestPickFirstLeaf_HappyEyeballs_TF_ThenTimerFires`.